### ISSUE TITLE
Loading adapters in training mode produces incorrect results and randomly switches back to base model

### ISSUE THREAD

**Issue by sei-cabidi** [huggingface/peft#2438]
### System Info

peft=0.14.0
python=3.10.15
pytorch=2.5.1+cu124
OS=RHEL

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [x] An officially supported task in the `examples` folder
- [x] My own task or dataset (give details below)

### Reproduction

# Debugging PEFT

I've noticed that PEFT randomly does not want to work
when loading in adapters. I have noticed that when loading an adapter
by ```PeftModel.from_pretrained(net, path_to_adapter, is_trainable=True)``` 
with ```is_trainable=True```, the predictions randomly revert to those
of the base model."

```python
import peft
import torch
import matplotlib.pyplot as plt

from copy import deepcopy
from pathlib import Path
from peft import get_peft_model, LoraConfig, PeftModel
from typing import Any

print(peft.__version__) # 0.14.0

# ============== Setup ============== #

# Define a simple network
class Network(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.a = torch.nn.Linear(1, 1)
        self.b = torch.nn.Linear(1, 1)
    
    def forward(self, x: torch.Tensor) -> torch.Tensor:
        y = self.a(x)
        y = self.b(y)
        return y
    
# Load the network    
net = Network()
for name, param in net.named_parameters():
    param.requires_grad = False

# Define config targeting layer "a"
peft_config = LoraConfig(
    inference_mode=False,
    r=7,
    lora_alpha=1,
    lora_dropout=0.1,
    target_modules=["a"]
)

# Wrap model in PEFT
peft_model = get_peft_model(net, peft_config)
peft_model.print_trainable_parameters()

# ============== Train ============== #

# Set up basic loss function, optimizer, scheduler
optimizer_kwargs: dict[str, Any] = {
    'momentum': 0.9,
    'weight_decay': 0,
    'nesterov': True,
    'lr': 1e-2,
}
params = list(filter(lambda p: p.requires_grad, peft_model.parameters()))
optimizer = torch.optim.SGD(
    params,
    **optimizer_kwargs
)
num_epochs: int = 1
scheduler_kwargs: dict[str, Any] = {
    'T_max': num_epochs,
}
scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(
    optimizer,
    **scheduler_kwargs
)
criterion = torch.nn.MSELoss()

# Simple training function
def train(n: int):
    input = torch.ones((1,1)) * 5
    truth = torch.ones((1,1)) * 3
    for _ in range(n):
        prediction = peft_model(input)
        print(f"prediction: {prediction}")
        print(f"truth: {truth}")
        loss = criterion(prediction, truth)
        print(f"loss: {loss:.3f}")
        loss.backward()
        optimizer.step()

# Train for one epoch and save the adapter
train(1)
Path("./adapter_1").mkdir(exist_ok=True)
peft_model.save_pretrained("./adapter_1")

# ============== Load adapter ============== #

# Load adapter into model in inference mode
first_model = PeftModel.from_pretrained(
    deepcopy(net),
    Path("./adapter_1"),
    autocast_adapter_dtype=False
)
print(f"First model: {list(first_model.parameters())}\n")

# Load adapter into model in training mode
second_model = PeftModel.from_pretrained(
    deepcopy(net),
    Path("./adapter_1"),
    is_trainable=True,
    autocast_adapter_dtype=False
)
print(f"Second model: {list(second_model.parameters())}\n")

# ============== Generate results ============== #

# Set number of samples
M = 100

# Produce results from the original model
with first_model.disable_adapter():
    original_results = torch.tensor([first_model(torch.ones((1,1))) for _ in range(M)]).reshape((10,10))
print(original_results)

# Produce M results from the first adapter model
first_results = torch.tensor([first_model(torch.ones((1,1))) for _ in range(M)]).reshape((10,10))
print(first_results)

# Produce M results from the first adapter model in training mode
second_results = torch.tensor([second_model(torch.ones((1,1))) for _ in range(M)]).reshape((10,10))
print(second_results)

# ============== Plot results ============== #

# Plot side-by-side
plt.imshow(torch.column_stack([original_results, first_results, second_results]))
plt.colorbar()
plt.tight_layout()
```

The bug can be seen here. I have plotted the predictions of three models stacked in (10,10)
grids for visual clarity.

1) The original base model
2) The first adapter loaded in inference mode
3) The first adapter loaded in training mode

Indeed it can be seen that loading the adapter in training mode both:
a) randomly generates predictions from the base model
b) has slightly incorrect predictions compared to the adapter in inference mode

### Expected behavior

When loading the adapter, regardless of whether ```is_trainable=True``` or ```is_trainable=False```, I expect the adapter to generate identical predictions. 

Instead, we see that when the adapter is loaded with ```is_trainable=True```, predictions randomly revert back to those that would be generated by the base model.

<img width="1110" alt="Image" src="https://github.com/user-attachments/assets/04a28a32-0494-408e-969a-26f3aac56a5a" />

**Comment by sei-cabidi** [comment]
I have also reproduced this behavior in PEFT 0.13.0, 0.13.1, 0.13.2 with a couple different version of PyTorch. This leads me to believe I'm either doing something very wrong, or there is a very subtle bug in PEFT.

**Comment by BenjaminBossan** [comment]
I haven't checked the details, but is this not most likely caused by the use of dropout being active in training mode? For such a small model, there is a certain chance that with dropout, the base model outputs are generated. Could you please try again with dropout set to 0?

**Comment by sei-cabidi** [comment]
That seemed to work!

Why would dropout cause a slight (consistent) change in prediction values?

**Comment by BenjaminBossan** [comment]
If all LoRA units are dropped out, which will happen with a certain probability, then only the base model results remain. The bigger the model and the lower the dropout rate, the less likely this is to happen.
